### PR TITLE
chore: refactor code quality issues

### PR DESCRIPTION
**Pull request by akshgpt7**
### Description
Hey, I'm a Developer Outreach at [DeepSource](https://deepsource.io/) and ran DeepSource analysis on my fork of the repo. It found some interesting [code quality and performance improvements](https://deepsource.io/gh/akshgpt7/ignite/) to consider.

This PR fixes a few of the issues detected for you to assess if it is right for you.
Happy to provide the tweaks separately otherwise :)

### Summary of changes

- Remove usage of `len()` for comparison.
- Remove reimported modules.
- Merge `isinstance` calls.
- Iterate dictionary directly instead of using `.keys()`.
- Remove unnecessary lambda expression.
- Add `.deepsource.toml` file for continuous analysis on bug risks/performance/code-quality issues on new changes.

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by vfdev-5** [comment]
@akshgpt7 thanks for the PR !  Could you please fix the conflict with master branch.

> Add .deepsource.toml file for continuous analysis on bug risks/performance/code-quality issues on new changes.

can you detail what it will exactly do ?

**Comment by akshgpt7** [comment]
@vfdev-5 The code-style test seems to fail after pulling from `master`.

> can you detail what it will exactly do ?

DeepSource is a static analysis tool. It'll detect bug-risks, performance issues, anti-patterns and some security vulnerabilities on each change and prompt them in the PR, just like other checks that run on a PR.
It'll significantly reduce code reviewing time and ensure code quality.
I'd share the quick steps to activate DeepSource if you'd like to activate it for this repository.

**Comment by KickItLikeShika** [comment]
@akshgpt7 Please read this section https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#code-development.
You have to use `isort`, `black`, `flake8` to solve the code-style issue.

**Comment by akshgpt7** [comment]
The tests are failing for `test_native.py`,  in which I have made no changes.

**Comment by vfdev-5** [comment]
@akshgpt7 yes, currently the master is broken, unfortunately.

If I understand correctly DeepSource is not activated now, in this case we do not need `.deepsource.toml` file.

> I'd share the quick steps to activate DeepSource if you'd like to activate it for this repository.

Can you link some written info on that. Thanks

**Comment by akshgpt7** [comment]
@vfdev-5 If you like, you can activate analysis for your repository to detect such code quality issues/bug risks on the fly for every change made. You can also use the Autofix feature to fix them with one click.

Here's what you can do if you wish to activate DeepSource to continuously analyze your repository:
- [Sign up](https://deepsource.io/signup) on DeepSource and activate analysis for this repository.
- Create `.deepsource.toml` configuration which you can use to configure your analysis settings (this PR already added that, but feel free to edit it anytime, or let me know if still want it removed from this PR since that is needed to configure the analysis).
- Track/Check analysis [here](https://deepsource.io/gh/pytorch/ignite).

If you have any doubts or questions, you can check out the [docs](https://deepsource.io/docs/), or feel free to reach out :)

**Comment by akshgpt7** [comment]
@vfdev-5 any updates on this?

**Comment by vfdev-5** [comment]
@akshgpt7 sorry for delay! This looks interesting in general and some of other project I know (e.g. MONAI) are using deepsource.
I'm a bit curious about how to compare deepsource to github/codeql which used by some other project I also know for example ?

btw, could you please resolve the conflict again.
Thanks!

**Comment by akshgpt7** [comment]
@vfdev-5 CodeQL is pretty different than DeepSource. It's an actual query language where you have to yourself write rules that are checked on code analysis.
DeepSource has pre-written rules that are checked to save you the hassle of writing rules for every check.

**Comment by akshgpt7** [comment]
@vfdev-5 fixed the conflicts :)

File: tests/ignite/engine/test_engine.py [hunk 1]
```diff
@@ -701,9 +701,6 @@ def restart_iter():
 
 
 def test_faq_inf_iterator_with_epoch_length():
-    # Code snippet from FAQ
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:704]: Let's keep this comment here.

File: tests/ignite/engine/test_engine.py [hunk 2]
```diff
@@ -726,9 +723,6 @@ def train_step(trainer, batch):
 
 
 def test_faq_inf_iterator_no_epoch_length():
-    # Code snippet from FAQ
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:729]: Same here.

File: tests/ignite/engine/test_engine.py [hunk 3]
```diff
@@ -755,9 +749,6 @@ def stop_training():
 
 
 def test_faq_fin_iterator_unknw_size():
-    # Code snippet from FAQ
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:758]: Here

File: tests/ignite/engine/test_engine.py [hunk 4]
```diff
@@ -782,10 +773,6 @@ def restart_iter():
     assert trainer.state.epoch == 5
     assert trainer.state.iteration == 5 * 11
 
-    # # # # #
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:785]: This one as well

File: tests/ignite/engine/test_engine.py [hunk 5]
```diff
@@ -807,9 +794,6 @@ def val_step(evaluator, batch):
 
 
 def test_faq_fin_iterator():
-    # Code snippet from FAQ
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:810]: Keep this

File: tests/ignite/engine/test_engine.py [hunk 6]
```diff
@@ -836,10 +820,6 @@ def restart_iter():
     assert trainer.state.epoch == 5
     assert trainer.state.iteration == 5 * size
 
-    # # # # #
```
Comment (vfdev-5) [inline tests/ignite/engine/test_engine.py:839]: Same here

**Comment by vfdev-5** [comment]
@akshgpt7 OK we can give a try to DeepSource and will see if it helps or bothers...
Let's fix final conflict here and we can merge it.

**Pull request closed without merging**